packages/filesystem/src/node/node-filesystem.ts

```diff
@@ -356,6 +353,40 @@ export class FileSystemNode implements FileSystem {
         return children;
     }
 
+    /** Return true if it's possible for this URI to have children.
+     *  It might not be possible to be certain because of permission problems
+     *  Or other filesystem errors.
+     */
+    protected mayHaveChildren(uri: URI): boolean {
+        /* If there's a problem reading the root directory.
+           Assume it's not empty to avoid overwriting anything.  */
+        try {
+            const rootStat = this.doGetStat(uri, 0);
+            if (rootStat === undefined) {
+                return true;
+            }
+            /* Not a directory.  */
+            if (rootStat !== undefined && rootStat.isDirectory === false) {
+                return false;
+            }
+        } catch {
+            return true;
+        }
+
+        /* If there's a problem with it's children then the directory must
+        not be empty.  */
+        try {
+            const stat = this.doGetStat(uri, 1);
```

Comment:
Between this and the above-mentioned `/* Not a directory.  */` check, someone might have deleted the file, hence `stat` can be `undefined`. So it is not necessarily true, but I do not think that previous logic was not atomic either. I would leave as is.